addons/pos_self_order/controllers/ordering.py

```diff
@@ -0,0 +1,455 @@
+        self, pos_config_sudo: PosConfig, table_access_token: str
+    ) -> Dict[str, Union[str, int]]:
+        """
+        :param pos_config_id: the id of the pos config for which we want to create the order
+        :param table_access_token: the access token of the table for which we want to create the order
+        :return: a dictionary containing the data that we need to create a new order
+        """
+        table_sudo = PosSelfOrderUtils._get_table_sudo(self, table_access_token)
+        pos_session_sudo = pos_config_sudo.current_session_id
+        if not table_sudo or not pos_session_sudo:
+            raise werkzeug.exceptions.Unauthorized()
+
+        sequence_number = self._get_sequence_number(table_sudo.id, pos_session_sudo.id)
+
+        return {
+            "id": self._generate_unique_id(pos_session_sudo.id, table_sudo.id, sequence_number),
+            "sequence_number": sequence_number,
+            "access_token": uuid.uuid4().hex,
+            "session_id": pos_session_sudo.id,
+            "table_id": table_sudo.id,
+        }
+
+    def _update_cart(
+        self,
+        cart: List[Dict],
+        existing_orderlines: List[Dict],
+    ) -> List[Dict]:
+        """
+        If the customer has an existing order, we will add the items from the existing order to the current cart.
+        (This is because the create_from_ui method will overwrite the old items from the order)
+        :param cart: The cart from the frontend.
+        :return: list of dictionaries with the items from the cart and the items from the existing order
+        """
+
+        return self._get_updated_orderlines(
+            existing_orderlines,
+            [item for item in cart if self._is_in_cart(item, existing_orderlines)],
+        ) + [item for item in cart if not self._is_in_cart(item, existing_orderlines)]
+
+    def _get_updated_orderlines(self, orderlines: List[Dict], cart: List[Dict]) -> List[Dict]:
+        """
+        :returns updated version of orderlines that takes into account the new items from the cart
+        """
+        return [
+            self._get_updated_orderline(line, updated_item)
+            if (
+                updated_item := next(
+                    (item for item in cart if self._is_same_product(item, line)), None
+                )
+            )
+            else line
+            for line in orderlines
+        ]
+
+    def _get_updated_orderline(self, line: Dict, updated_item: Dict) -> Dict:
+        """
+        This function assumes that the price of the product was not changed between the time the
+        first items were added to the order and the time the new items were added to the order.
+        """
+        qty_ratio = (updated_item["qty"] + line["qty"]) / line["qty"]
+        return {
+            **line,
+            **{
+                "price_subtotal": line["price_subtotal"] * qty_ratio,
+                "price_subtotal_incl": line["price_subtotal_incl"] * qty_ratio,
```

Comment:
This calculation of "new total prices" after quantity update is wrong. Should always trigger the whole orderline price calculation because unit prices can vary depending on qty because of pricelists. (That is if pricelists are still there. I remember a conversation that they're trying to abolish it.)

Reply:
is it ok now?